src/ucp/tag/offload.h

```diff
@@ -85,4 +88,53 @@ ucp_tag_offload_try_cancel(ucp_worker_t *worker, ucp_request_t *req, int force)
     }
 }
 
+/* For homogeneous clusters it is expected that tag offload messages will arrive
+ * from the single interface. Otherwise need to add a particular iface to tags
+ * hash, where key is a tag masked with tag_sender mask. */
+static UCS_F_ALWAYS_INLINE void
+ucp_tag_offload_unexp(ucp_worker_iface_t *iface, ucp_tag_t tag)
+{
+    ucp_worker_t *worker   = iface->worker;
+    ucp_context_t *context = worker->context;
+    ucp_tag_t tag_key;
+    khiter_t hash_it;
+    int ret;
+
+    if (ucs_unlikely(!context->config.ext.tm_offload)) {
+        /* Can happen when different processes have different configuration,
+         * which is very uncommon. */
+        return;
+    }
+
+    if (ucs_unlikely(!(iface->flags & UCP_IFACE_FLAG_OFFLOAD_ENABLED))) {
+        if (ucs_queue_is_empty(&worker->tm.offload.ifaces)) {
+            ucs_assert(worker->tm.offload.thresh       == SIZE_MAX);
+            ucs_assert(worker->tm.offload.zcopy_thresh == SIZE_MAX);
+
+            worker->tm.offload.thresh       = context->config.ext.tm_thresh;
+            worker->tm.offload.zcopy_thresh = context->config.ext.tm_max_bcopy;
+
+            ucs_debug("Enable TM offload: thresh %zu, zcopy_thresh %zu",
+                    worker->tm.offload.thresh, worker->tm.offload.zcopy_thresh);
+        }
+
+        ucp_worker_iface_activate(iface, 0);
```

Comment:
need to somehow unite this with the generic interface activation mechanism, when iface_progress is called explicitly when getting interrupt, and then counting active messages on the interface. i guess need to add option to modify unexp callback function after interface is already created, same as for active messages